## One Hot Encoding - variables with many categories

In [1]:
import pandas as pd
import numpy as np

In [4]:
data = pd.read_csv("mercedesbenz.csv" ,usecols=["X1","X2","X3","X4","X5"])

In [5]:
data.head()

X1  X2 X3 X4 X5
0  v  at  a  d  u
1  t  av  e  d  y
2  w   n  c  d  x
3  t   n  f  d  x
4  v   n  f  d  h

In [12]:
##lets have a look at how many labels each varible has
for col in data.columns:
    print(col ,':',len(data[col].unique()) , "labels")

X1 : 27 labels
X2 : 44 labels
X3 : 7 labels
X4 : 4 labels
X5 : 29 labels


In [13]:
##let's examine how many columns we will obtain after one hot encoding these varibles
pd.get_dummies(data,drop_first = True).shape

(4209, 106)

we can see that from just 6 intial categorical variables,  we end up with 106 new varibles.


In [16]:
##let's find the top 10 most frequet categories for varible x2

data.X2.value_counts().sort_values(ascending = False).head(20)

as    1659
ae     496
ai     415
m      367
ak     265
r      153
n      137
s       94
f       87
e       81
aq      63
ay      54
a       47
t       29
k       25
i       25
b       21
ao      20
ag      19
z       19
Name: X2, dtype: int64

In [19]:
##let's make a list with the most frequent frequent categories of the varible
top_10 = [X for X in data.X2.value_counts().sort_values(ascending = False).head(20).index]

In [21]:
top_10

['as',
 'ae',
 'ai',
 'm',
 'ak',
 'r',
 'n',
 's',
 'f',
 'e',
 'aq',
 'ay',
 'a',
 't',
 'k',
 'i',
 'b',
 'ao',
 'ag',
 'z']

In [22]:
for label in top_10:
    data[label]= np.where(data['X2'] == label , 1 , 0)
    
data[['X2']+top_10].head(40)    
    

X2  as  ae  ai  m  ak  r  n  s  f  ...  aq  ay  a  t  k  i  b  ao  ag  z
0   at   0   0   0  0   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
1   av   0   0   0  0   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
2    n   0   0   0  0   0  0  1  0  0  ...   0   0  0  0  0  0  0   0   0  0
3    n   0   0   0  0   0  0  1  0  0  ...   0   0  0  0  0  0  0   0   0  0
4    n   0   0   0  0   0  0  1  0  0  ...   0   0  0  0  0  0  0   0   0  0
5    e   0   0   0  0   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
6    e   0   0   0  0   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
7   as   1   0   0  0   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
8   as   1   0   0  0   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
9   aq   0   0   0  0   0  0  0  0  0  ...   1   0  0  0  0  0  0   0   0  0
10   r   0   0   0  0   0  1  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
11   e   0   0   0  0   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
12  ai   0   0   1  0   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
13   e   0   0   0  0   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
14  as   1   0   0  0   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
15  ak   0   0   0  0   1  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
16   e   0   0   0  0   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
17  as   1   0   0  0   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
18   m   0   0   0  1   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
19  ak   0   0   0  0   1  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
20   m   0   0   0  1   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
21  as   1   0   0  0   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
22   a   0   0   0  0   0  0  0  0  0  ...   0   0  1  0  0  0  0   0   0  0
23   m   0   0   0  1   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
24  ak   0   0   0  0   1  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
25   m   0   0   0  1   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
26  as   1   0   0  0   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
27  as   1   0   0  0   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
28  as   1   0   0  0   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
29   m   0   0   0  1   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
30  as   1   0   0  0   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
31  as   1   0   0  0   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
32   m   0   0   0  1   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
33  as   1   0   0  0   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
34  as   1   0   0  0   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
35  ai   0   0   1  0   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
36  as   1   0   0  0   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
37   k   0   0   0  0   0  0  0  0  0  ...   0   0  0  0  1  0  0   0   0  0
38  as   1   0   0  0   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0
39  as   1   0   0  0   0  0  0  0  0  ...   0   0  0  0  0  0  0   0   0  0

[40 rows x 21 columns]

In [24]:
##get whole set of dummy varibles, for all the categorical variables

def one_hot_topx(df,variable,top_x_labels):
    ##fuction to create the dumy varibles for the most frequent labels
    ## we can vary the number of most frequent label that we encode
    
    for label in top_x_labels:
        df[variable + "-"+label]=np.where(data[variable]==label,1,0)
        
##read the data again
data = pd.read_csv("mercedesbenz.csv",usecols = ["X1","X2","X3","X4","X5","X6"])

##encode x2 into 10 most frequqent categoris
one_hot_topx(data,"X2",top_10)
data.head()

X1  X2 X3 X4 X5 X6  X2-as  X2-ae  X2-ai  X2-m  ...  X2-aq  X2-ay  X2-a  \
0  v  at  a  d  u  j      0      0      0     0  ...      0      0     0   
1  t  av  e  d  y  l      0      0      0     0  ...      0      0     0   
2  w   n  c  d  x  j      0      0      0     0  ...      0      0     0   
3  t   n  f  d  x  l      0      0      0     0  ...      0      0     0   
4  v   n  f  d  h  d      0      0      0     0  ...      0      0     0   

   X2-t  X2-k  X2-i  X2-b  X2-ao  X2-ag  X2-z  
0     0     0     0     0      0      0     0  
1     0     0     0     0      0      0     0  
2     0     0     0     0      0      0     0  
3     0     0     0     0      0      0     0  
4     0     0     0     0      0      0     0  

[5 rows x 26 columns]

In [25]:
top_10 = [X for X in data.X1.value_counts().sort_values(ascending = False).head(20).index]
one_hot_topx(data,"X1",top_10)
data.head()

X1  X2 X3 X4 X5 X6  X2-as  X2-ae  X2-ai  X2-m  ...  X1-w  X1-z  X1-u  X1-e  \
0  v  at  a  d  u  j      0      0      0     0  ...     0     0     0     0   
1  t  av  e  d  y  l      0      0      0     0  ...     0     0     0     0   
2  w   n  c  d  x  j      0      0      0     0  ...     1     0     0     0   
3  t   n  f  d  x  l      0      0      0     0  ...     0     0     0     0   
4  v   n  f  d  h  d      0      0      0     0  ...     0     0     0     0   

   X1-m  X1-t  X1-h  X1-y  X1-f  X1-j  
0     0     0     0     0     0     0  
1     0     1     0     0     0     0  
2     0     0     0     0     0     0  
3     0     1     0     0     0     0  
4     0     0     0     0     0     0  

[5 rows x 46 columns]

In [26]:
top_10 = [X for X in data.X3.value_counts().sort_values(ascending = False).head(20).index]
one_hot_topx(data,"X3",top_10)
data.head()

X1  X2 X3 X4 X5 X6  X2-as  X2-ae  X2-ai  X2-m  ...  X1-y  X1-f  X1-j  X3-c  \
0  v  at  a  d  u  j      0      0      0     0  ...     0     0     0     0   
1  t  av  e  d  y  l      0      0      0     0  ...     0     0     0     0   
2  w   n  c  d  x  j      0      0      0     0  ...     0     0     0     1   
3  t   n  f  d  x  l      0      0      0     0  ...     0     0     0     0   
4  v   n  f  d  h  d      0      0      0     0  ...     0     0     0     0   

   X3-f  X3-a  X3-d  X3-g  X3-e  X3-b  
0     0     1     0     0     0     0  
1     0     0     0     0     1     0  
2     0     0     0     0     0     0  
3     1     0     0     0     0     0  
4     1     0     0     0     0     0  

[5 rows x 53 columns]

In [27]:
top_10 = [X for X in data.X4.value_counts().sort_values(ascending = False).head(20).index]
one_hot_topx(data,"X4",top_10)
data.head()

X1  X2 X3 X4 X5 X6  X2-as  X2-ae  X2-ai  X2-m  ...  X3-f  X3-a  X3-d  X3-g  \
0  v  at  a  d  u  j      0      0      0     0  ...     0     1     0     0   
1  t  av  e  d  y  l      0      0      0     0  ...     0     0     0     0   
2  w   n  c  d  x  j      0      0      0     0  ...     0     0     0     0   
3  t   n  f  d  x  l      0      0      0     0  ...     1     0     0     0   
4  v   n  f  d  h  d      0      0      0     0  ...     1     0     0     0   

   X3-e  X3-b  X4-d  X4-a  X4-c  X4-b  
0     0     0     1     0     0     0  
1     1     0     1     0     0     0  
2     0     0     1     0     0     0  
3     0     0     1     0     0     0  
4     0     0     1     0     0     0  

[5 rows x 57 columns]

In [28]:
top_10 = [X for X in data.X5.value_counts().sort_values(ascending = False).head(20).index]
one_hot_topx(data,"X5",top_10)
data.head()

X1  X2 X3 X4 X5 X6  X2-as  X2-ae  X2-ai  X2-m  ...  X5-ae  X5-ag  X5-ac  \
0  v  at  a  d  u  j      0      0      0     0  ...      0      0      0   
1  t  av  e  d  y  l      0      0      0     0  ...      0      0      0   
2  w   n  c  d  x  j      0      0      0     0  ...      0      0      0   
3  t   n  f  d  x  l      0      0      0     0  ...      0      0      0   
4  v   n  f  d  h  d      0      0      0     0  ...      0      0      0   

   X5-ab  X5-l  X5-af  X5-ad  X5-k  X5-c  X5-j  
0      0     0      0      0     0     0     0  
1      0     0      0      0     0     0     0  
2      0     0      0      0     0     0     0  
3      0     0      0      0     0     0     0  
4      0     0      0      0     0     0     0  

[5 rows x 77 columns]